### ISSUE TITLE
Iframe accesibile instance of sim that displays screen reader output

### ISSUE THREAD

**Issue by ariel-phet** [phetsims/balloons-and-static-electricity#134]
@samreid this seems like a good question for you...

@emily-phet was curious if it would be possible to embed the accessible instance of BASE in an iframe and below it have a text box that displays the screen-reader output just as printed text as you interact with sim. 

Thoughts on if this is simple/feasible? Would it be possible to whip out a quick and dirty demo?

**Comment by samreid** [comment]
It is a question for Jesse about how easy it is to get the screen reader text programmatically.  @jessegreenberg can you comment?

**Comment by samreid** [comment]
@ariel-phet what is the timeline for this?  @jessegreenberg can you comment on accessing the text?  Also, this is not something that would necessarily need phet-io, you can deliver the strings over postMessage() quite easily.

**Comment by ariel-phet** [comment]
It is basically something that would be very nice for demos. Also potentially very useful if we put up an example version, and allow the accessibility strings to be translatable (this way a translator could see the printed out strings without having to use a screen reader). Worth some investigation to see if we can make a very barebones solution. Perhaps @samreid and @jessegreenberg could spend an hour in the next couple of weeks seeing if you can whip up something quick and dirty.

**Comment by jessegreenberg** [comment]
Unfortunately this is not possible because screen readers do not provide a way to access their output programatically.

It is also very difficult to predict what a screen reader *might* say because it works with a virtual copy of our parallel DOM.  The browser has no way of knowing where the user's cursor is in the virtual buffer.

NVDA and JAWS both keep logs of their output, but these are meant for development purposes.  They cannot be used as a way to get and manipulate the output of the screen reader in real time.

Many of our accessibility strings can only be found if AT is in use and the user is familiar with common screen reader navigation strategies (such as 'h' for headings and arrow keys to navigate text content).  We will need another way for translators to review translations of accessible content.  Maybe a visually friendly render of the parallel DOM shown to the side of the sim? Not sure if that would make sense to the translator and would take some design work.

Back to @ariel-phet for further discussion.

**Comment by emily-phet** [comment]
@jessegreenberg: I didn't realize it would be so difficult to access the text for different purposes. My understanding was that the text exists in the DOM, and is tagged for screenreader to access...is it the case that as structured *nothing else* can access this text? I figured since it's sitting there in the DOM, it could be accessed in other ways. I can see the challenge (or impossibility, as you say above) for the non-screenreader "thing" to access the text and output it simultaneously with the screenreader.

What if you imagine trying to couple auditory descriptions with basic keyboard navigation (no screen reader involved). Would this be possible? (not that we should do it, I'm just curious about what might be different about this scenario.)

**Comment by samreid** [comment]
@jessegreenberg perhaps we could investigate using `document.activeElement` to identify the active DOM element in the accessible (parallel) DOM, then based on its type, getting the text from the element.

**Comment by samreid** [comment]
Here's a possible way to detect focus changes: http://stackoverflow.com/questions/10035564/is-there-a-cross-browser-solution-for-monitoring-when-the-document-activeelement

**Comment by jobara** [comment]
Screenreaders such as JAWS and NVDA use a virtual copy of the Document that a user can navigate through. When a user interacts with the content in this way, the browser is unaware of what content is currently being accessed by the user. I believe this post was shared on another thread earlier, but I'll post it here again just in case http://tink.uk/understanding-screen-reader-interaction-modes/ .

In terms of accessing aria attributes and other markup from the parallel DOM, perhaps you could look at a tool like [tota11y](http://khan.github.io/tota11y/); which includes an experimental screen reader wand. It's not perfect and may not have the details you want, but could give you an idea of how to make something, or perhaps you could tweak their implementation. The trick in this case may be informing tota11y which parallel DOM element is associated with which sim element. Also I believe just outputting the aria attributes and other content from the parallel DOM to some other portion of the UI should be possible, assuming:

* the output occurs with browser events (e.g. focus, key events, etc.)
* the output occurs with events from the sim (e.g. model changes, etc.)
* the output doesn't have to be the same as an AT's 
  * JAWS and NVDA don't announce things the same, the best we could do is a simulation or the bare text

**Comment by ariel-phet** [comment]
@samreid - although certainly not a simple solution...if a sim were instrumented for PhET-iO I would think there would be a way to take the normal scenery input events and such and play them back through the parallel DOM (though it sounds like this would take some sort of custom interpreter or such).

@jessegreenberg I would still like you and @samreid to take an hour or two and do a brainstorming/hack session at some point in the next week or two.

**Comment by jessegreenberg** [comment]
> My understanding was that the text exists in the DOM, and is tagged for screenreader to access...

@emily-phet, this understanding is correct, and we can access and use this text in any way we wish.  The challenge here is that we have less control over the screen reader.  It is difficult to get and show what is actively being read.

> What if you imagine trying to couple auditory descriptions with basic keyboard navigation (no screen reader involved). Would this be possible?

Yes, this would be much easier without screen readers. With another text to speech solution, we could have direct control of the output and a way to get what is being read.

**Comment by jessegreenberg** [comment]
> perhaps we could investigate using document.activeElement to identify the active DOM element in the accessible (parallel) DOM, then based on its type, getting the text from the element.

For the reasons that @jobara mentioned in https://github.com/phetsims/balloons-and-static-electricity/issues/134#issuecomment-199320798 ```document.activeElement``` is probably not enough to get all of our descriptions.  When navigating with the virtual cursor, the screen reader does not set focus until it reaches a form element or an element with tabIndex=0, which excludes many of our labels and descriptions.

**Comment by jessegreenberg** [comment]
@jobara, that tool for modeling the virtual cursor looks really great!  Before something like that is possible, we will likely need to complete https://github.com/phetsims/scenery/issues/41.  A tool that works like that could be very nice for reviewing translated strings and accessible content.

> Also I believe just outputting the aria attributes and other content from the parallel DOM to some other portion of the UI should be possible, assuming:
 - the output occurs with browser events (e.g. focus, key events, etc.)
 - the output occurs with events from the sim (e.g. model changes, etc.)
 - the output doesn't have to be the same as an AT's
  JAWS and NVDA don't announce things the same, the best we could do is a simulation or the bare text

This will still be challenging because we can't count on some of these assumptions.  The ```keydown``` event is usually not fired when the screen reader is active.  And ```focus``` only changes when the virtual cursor is over a form element.  Text is usually read without a model change.

**Comment by samreid** [comment]
@emily-phet & @ariel-phet are you intending that this iframe demo would be viewed with a screen reader on or off (or support for both)?

**Comment by ariel-phet** [comment]
Screen reader off most likely.

For instance, @emily-phet has commented that in interviews with students using a screen reader, the text is read so quickly, a normal user would not be able to interpret it.

Second, for use for our translators in the future, most will not have access to a screen reader or expertise on how to use one.

**Comment by jessegreenberg** [comment]
If we assume the screen reader is off, we can use the strategies @jobara mentioned in https://github.com/phetsims/balloons-and-static-electricity/issues/134#issuecomment-199320798 to listen for all of the necessary events.

**Comment by jessegreenberg** [comment]
In #538, @samreid and I started working on a prototype for this.  The solution uses a custom virtual cursor that can be controlled with the arrow keys to find all content and descriptions in the accessible DOM.  Here is the prototype:

http://www.colorado.edu/physics/phet/dev/html/balloons-and-static-electricity/1.2.0-accessible-instance.13/screen-reader.html

This version also has support for aria-live updates and queues the text if multiple updates happen at once. It includes a few other changes so that labels are not repeated.

This is very much a prototype, but should serve to show what this kind of tool could look like and be used for.

**Comment by jessegreenberg** [comment]
Closing this issue.  We have made progress on this prototype.  We also have other prototypes that could be useful in solving the issue described in the original ticket and in https://github.com/phetsims/balloons-and-static-electricity/issues/134#issuecomment-199416485.

For instance, see 
http://www.colorado.edu/physics/phet/dev/html/balloons-and-static-electricity/1.2.0-reader.6/balloons-and-static-electricity_en.html?reader

and
http://www.colorado.edu/physics/phet/dev/html/balloons-and-static-electricity/1.2.0-reader.7/acccessibility-test.html
